src/tachimawari/control/packet/protocol_1/status/bulk_read_data.cpp

```diff
@@ -0,0 +1,186 @@
+    if (rxpacket->at(header_place) == 0xFF && rxpacket->at(header_place + 1) == 0xFF) {
+      break;
+    } else if (header_place == (packet_length - 2) && rxpacket->at(packet_length - 1) == 0xFF) {
+      break;
+    }
+  }
+
+  if (header_place == 0) {
+    return packet_length;
+  } else {
+    for (size_t i = 0; i < (packet_length - header_place); ++i) {
+      rxpacket->at(i) = rxpacket->at(i + header_place);
+    }
+
+    return packet_length - header_place;
+  }
+}
+
+int BulkReadData::update_all(
+  std::shared_ptr<std::map<uint8_t, BulkReadData>> bulk_data,
+  std::shared_ptr<std::vector<uint8_t>> rxpacket, int packet_length, int data_number)
+{
+  while (true) {
+    int header_place = 0;
+    for (header_place = 0; header_place < (packet_length - 1); ++header_place) {
+      if (rxpacket->at(header_place) == 0xFF && rxpacket->at(header_place + 1) == 0xFF) {
+        break;
+      }
+    }
+
+    if (header_place == 0) {
+      uint8_t packet_id = rxpacket->at(PacketIndex::ID);
+
+      if (bulk_data->find(packet_id) != bulk_data->end()) {
+        int length = rxpacket->at(PacketIndex::LENGTH) + 4;
+
+        if (bulk_data->at(packet_id).is_valid(*rxpacket.get())) {
+          int bulk_data_length = rxpacket->at(PacketIndex::LENGTH) + 4;
+
+          for (size_t i = 0; i < packet_length - bulk_data_length; ++i) {
+            rxpacket->at(i) = rxpacket->at(i + bulk_data_length);
+          }
+
+          data_number--;
+          packet_length -= bulk_data_length;
+        } else {
+          // so RX_CORRUPT
+          for (size_t i = 0; i < packet_length - 2; ++i) {
+            rxpacket->at(i) = rxpacket->at(i + 2);
+          }
+
+          packet_length -= 2;
+        }
+      }
+
+      if (data_number <= 0 || packet_length <= 0) {
+        return data_number;
+      }
+    } else {
+      for (size_t i = 0; i < (packet_length - header_place); ++i) {
+        rxpacket->at(i) = rxpacket->at(i + header_place);
+      }
+
+      packet_length -= header_place;
+    }
+  }
+}
+
+BulkReadData::BulkReadData(
+  uint8_t id, uint8_t starting_address,
+  uint8_t data_length)
+: Packet(id, 0xFF), start_address(starting_address), data_length(data_length)
+{
+}
+
+bool BulkReadData::is_valid(std::vector<uint8_t> rxpacket)
+{
+  info = rxpacket[PacketIndex::ERROR];
+
+  int packet_length = rxpacket[PacketIndex::LENGTH] + 4;
+  for (size_t i = PacketIndex::PARAMETER; i < packet_length - 1; ++i) {
+    parameters.push_back(rxpacket[i]);
+  }
+
+  calculate_checksum();
+  if (checksum == rxpacket[packet_length - 1]) {
+    return true;
+  } else {
+    info = 0xFF;
+    parameters.clear();
+    checksum = 0x00;
+
+    return false;
+  }
+}
+
+bool BulkReadData::is_filled() const
+{
+  return parameters.size() != 0;
+}
+
+int BulkReadData::get(uint8_t address) const
+{
+  if (address >= start_address) {
+    size_t index = static_cast<size_t>(address - start_address);
+
+    if (index < data_length) {
+      return parameters[index];
+    }
+  }
```

Comment:
what if `start_address` is negative? i see that it's signed, so there's a possibility that it could be negative

Reply:
Yeah, it's signed integer, however if we see at the source where it is initialized, the variable is initialized only by `uint8_t`. To make it clear, maybe I will change the data type to `uint8_t`.